### DISCUSSION TITLE
Proprietary USB Input Device Emulator

### DISCUSSION THREAD

**Discussion by 641i130** [hathach/tinyusb#1410] [Q&A]
I'm trying to build a USB device emulator using this library on the RP2040. I'm kind of lost on how/where to start. I figured out how to get the ids to match but the bInterfaceClass is still set to HID. I just modified the code from this repo (https://github.com/Drewol/rp2040-gamecon). I want to set it to 255. I'm not sure where to find good documentation/guides/examples on building a generic input device for USB that isn't cluttered with HID or xinput stuff. I have the endpoint information as well. Any input to this project I'm starting is appreciated!

Some other notes:
This device is an arcade machine USB input device that I'm trying to develop a way to interface with without modifying the game software at all.

**Comment by hathach** [comment]
what is your particular issue though ? If you want to learn USB, USB complete is a good book.

**Reply by 641i130** [reply]
After a few ddgs I think the proper terminology for what I'm struggling with is making a custom descriptor and endpoint. I couldn't find any good examples on how to do this.
I'll look into that book, it looks very useful!

**Reply by hathach** [reply]
which descriptor specifically , configuration or hid report descriptor ? best way to learn is from example. Which OS you are on windows/linux/macos ? I could give some hint on how to obtain those descriptor from existing device so that you could try to mimic

**Reply by 641i130** [reply]
![image](https://user-images.githubusercontent.com/46945263/160653709-bb733a3b-5a43-43b8-8b5e-4d363d5a1281.png)
Here's the `lsusb` output on Linux, left is the RP2040 I'm trying to develop this on. I just changed the descriptors of an HID gamepad example I found. The right has the actual device information. Everything above these lines is matching (like product id and vendor id).

**Reply by hathach** [reply]
that you already have an example, if you are trying to make an gamepad, have you looked at `hid_composite` example and its descriptor. Maybe try to run to see what the example does.

https://github.com/hathach/tinyusb/blob/master/examples/device/hid_composite/src/usb_descriptors.c#L82

**Reply by 641i130** [reply]
The thing is, it's not an HID device at all, at least to my understanding based on the `lsusb`. I'm lost on how to implement custom USB reports I guess if that's the right terminology. Making a custom USB class is the right way to go right? I'm just unsure.

**Reply by hathach** [reply]
You should have specified this in the 1st post. For vendor, it is all vendor, from firmware, data to pc driver. There is nothing special, everything is up to you to decide how it work.

**Reply by 641i130** [reply]
Sorry, I thought I wrote that somewhere. So where do I start writing a vendor-specific class in terms of TinyUSB? Are there such examples yet?

**Reply by hathach** [reply]
`webusb_serial` is the one that make use of vendor interface.
